docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,298 @@
+    samples = cdf_sampler.sample(size=100)
+
+The returned object is a `np.ndarray` containing the sampled indices.
+
+In addition the `InverseCDFSampler` should have the possibility to sample only 
+along a given axis of the PDF. This features is required to handle sampling of 
+multiple PDFs at the same time to avoid a Python loop over the PDFs. This will 
+be supported by specifying the `axis` argument:
+
+.. code::
+
+    cdf_sampler = InverseCDFSampler(pdf=pdfs, axis=0)
+    samples = cdf_sampler.sample(size=1)
+
+
+The `InverseCDFSampler` could addtionally implement a few helper functions to plot
+the CDF and PDF. The `InverseCDFSampler` class will replace the current
+`GeneralRandom` and `GeneralRandomArray` classes in `gammapy.utils.distributions`. 
+
+
+
+MapEventSampler
+---------------
+
+To support sampling from the predicted number of counts maps for source as well 
+as background maps, we propose to introduce a `MapEventSampler` class. The 
+responsibility of this class is to handle the physical coordinate transformations, 
+time dependence of the model, output data structures and sampling of the total 
+number of observed events.
+Starting from a predicted number of counts map (e.g. computed by the `MapEvaluator`):
+
+.. code::
+
+    src = SkyModel()
+    evaluator = MapEvaluator(src, exposure, psf=None, edisp=None)
+    npred_map = evaluator.compute_npred()
+
+The `MapEventSampler` takes this map and a lightcurve model as an input and 
+allows to sample from it with the `.sample()` method:
+
+.. code::
+
+    lightcurve = LightCurveTableModel() # or PhaseCurveTableModel()
+    src_sampler = MapEventSampler(npred_map, lightcurve, t_min=, t_max=)
+
+    # draw total number of events
+    n_events = src_sampler.npred_total()
+
+    # draw events as an astropy table
+    events_src = map_sampler.sample()
+
+If no lightcurve is provided a constant rate can be assumed e.g. for 
+background models:
+
+.. code::
+
+    bkg_map = BackgroundModel().map
+    bkg_sampler = MapEventSampler(bkg_map, t_min=, t_max=)
+    events_bkg = bkg_sample.sample()
+
+
+The `MapEventSampler` returns an `astropy.table.Table` object, with the following
+columns: `TIME`, `RA_TRUE`, `DEC_TRUE`, `ENERGY_TRUE`. Later additional info such
+as `DETX` and `DETY` could be added.
+
+
+IRFEventDistributor
+-------------------
+
+We also propose a class to handle correctly the spatially varying IRFs: the
+`IRFEventDistibutor` class. Based on the `PSFMap` and `EDispMap`
+as an input it interpolates the "correct" IRF at the position of a given event and
+applies it. In more detail the class will calculate the psf and the energy dispersion 
+for each event true position and true energy. The IRFs are assumed to be constant
+and not time-dependent. The time dependency will be handled higher level, by a loop
+over good time-intervals (GTIs). The class would be used as following:
+
+.. code::
+
+    events = vstack([events_src_1, events_src_2, ..., events_src_N])
```

Comment:
Suggest to remove this vstack in the code example. See #2277 how controversial this event stacking / GTI handling is. We don't want to re-hash this discussion here.

It should work like this: EVENTS always have an associated GTI table. Often it's a single row and one interval, but it could be multiple intervals. The assumption is that IRFs and pointing are constant during all of those GTIs. You only need to consider GTI when sampling the event TIMEs.

Otherwise in your interface I see that you use `exposure` and assume it's pre-computed already. That's good. Note that it will be computed as `exposure = a_eff * total_time` where `total_time = sum(gti.stop - gti.start)`, i.e. it's a single exposure map that contains the exposure from all GTIs (often in simulations there will be one interval, but that's fine, it doesn't cost anything to support multiple intervals as long as the IRFs and pointing are the same).

So explain this where you explain how you sample `TIME`. You first sample one interval `[0, total_time]` and then re-map those times according to the GTI intervals. You can assume GTI are sorted for simulation. If you want, you can also just assume a single GTI interval for now for simulation, that's fine. But please be a bit more clear how `GTI` and `TIME` exactly will be handled.